kvdb/etcd/stm.go

```diff
@@ -320,42 +345,224 @@ func runSTM(s *stm, apply func(STM) error) error {
 	return executeErr
 }
 
-// add inserts a txn response to the read set. This is useful when the txn
-// fails due to conflict where the txn response can be used to prefetch
-// key/values.
-func (rs readSet) add(txnResp *v3.TxnResponse) {
-	for _, resp := range txnResp.Responses {
+func newReadSet() *readSet {
+	return &readSet{
+		tree:       btree.New(2),
+		fullRanges: make(map[string]bool),
+	}
+}
+
+// add inserts key/values to to read set.
+func (rs *readSet) add(responses []*pb.ResponseOp) {
+	for _, resp := range responses {
 		getResp := (*v3.GetResponse)(resp.GetResponseRange())
 		for _, kv := range getResp.Kvs {
-			rs[string(kv.Key)] = stmGet{
-				val: string(kv.Value),
-				rev: kv.ModRevision,
+			rs.addItem(
+				string(kv.Key), string(kv.Value), kv.ModRevision,
+			)
+		}
+	}
+}
+
+// addFullRange adds all full ranges to the read set.
+func (rs *readSet) addFullRange(prefixes []string, responses []*pb.ResponseOp) {
```

Comment:
Are we always sure these two slices are the same length?

Reply:
Yes this is ensured by etcd's request/response protocol.